chain/chain/src/stateless_validation/chunk_endorsement.rs

```diff
@@ -67,6 +77,7 @@ impl Chain {
             // Verify that the signature in block body are valid for given chunk_validator.
             // Signature can be either None, or Some(signature).
             // We calculate the stake of the chunk_validators for who we have the signature present.
+            // We always include block producers who have sent approvals, since that indicate they have applied the corresponding chunk
```

Comment:
I do not get this. My understanding is in stateless validation, block producers do not apply chunks but only validate blocks (headers), so their approval will not be for the application of chunks. Even if they are also chunk validators how do we relate their approval with witness validation?

Reply:
Below (L109-112) the approvals are filtered down to those from chunk producers for the shard. So if you are a chunk producer and also a block producer, your block approval is implictly an endorsement of the chunk. 

The part I'm not sure I understand is this: the ultimate goal is to "stop chunk producers from validating state witness if they are assigned also as a validator". Would it be simpler to just send an endorsement without validating in that case instead of also adding this implicit endorsement logic?

Reply:
ok thanks Saketh, I missed that part. Maybe change comment to
// We always include chunk producers that are also block producers who have sent approvals, since that indicate they have applied the corresponding chunk

---

This (chunk producer sending endorsement right away) is already happening here: https://github.com/near/nearcore/blob/7b68efbd9bdce6c97841422891866941704b1601/chain/client/src/stateless_validation/state_witness_producer.rs#L59
(the missing piece is that this does not prevent chunk producer to send witness to itself and validate it)

is the concern that that endorsement may not reach the right block producer (eg. in case of skips)?

Reply:
> the ultimate goal is to "stop chunk producers from validating state witness if they are assigned also as a validator"

This is only part of the goal. The other part is to allow chunk producers to use block approvals as chunk endorsements so that their chunk endorsements would arrive earlier, which increases the likelihood of a chunk getting enough endorsements and being included in a block.